Add recounting package.

zenhack opened the pull request:
Inspired by the way Client refcounting works in go-capnp.

Needs docs & tests still.

---

@lthibault, thinking through how to deal with the promise resolution stuff, I'm finding myself wanting to do more refcounting, and it seems worthwhile to not add another bespoke implementation -- so I decided to pull out the basic pattern and add it to go-util. This needs docs & tests, but is otherwise representative of what I think this should look like; thoughts?

zenhack committed:
Add recounting package.

Inspired by the way Client refcounting works in go-capnp.

Needs docs & tests still.

lthibault reviewed (approved):
I like it.  Is this going to be used for clients as well?

zenhack commented:
Yeah, it probably makes sense to have it replace the existing logic.

zenhack committed:
rc: Docs, delete interface.go.

I'm not quite sure how the interfaces fit in yet, and Releaser feels
wrong given Ref takes a separate callback. Easier to add than remove.

zenhack commented:
Alright, added docs & tests. I also deleted the interfaces, since I don't see myself using them immediately, I feel weird about the fact that NewRef doesn't use Releaser, and it's always easy to add it later.

zenhack committed:
rc: add tests.
